### ISSUE TITLE
Investigate 'editors can only edit their features' setting

### ISSUE THREAD

**Issue by allisonmuise** [Esri/crowdsource-manager#261]
currently, this setting doesn't appear to restrict editing features though manager. shouldn't this be enforced by the API?

**Comment by pratikg** [comment]
@allisonmuise,
 
> Just to clarify, this should only be the case when the feature layer has the 'Editors can only edit their features' option enabled.

Ok.
In the code, we will check the following condition and decide whether 'Editors can only edit their features' option is enabled or not.

**Condition:** We will check whether "ownershipBasedAccessControlForFeatures" property is available in the layer and it's not empty. If its empty, it means, "Editors can only edit their features" is not enabled.

![image](https://user-images.githubusercontent.com/1454356/49630169-fe716600-fa12-11e8-81a9-2925e9276e56.png)

> Could you also investigate if this is something the API should be enforcing? I don't recall that we've had to explicitly enforce this anywhere else.

We have investigated and found that the feature layer has a method "getEditCapabilities" to which selected feature needs to be passed as a "feature" parameter. In return, it provides an object which has a key "canUpdate" which states whether signed in user could edit the feature. Depending on this we will allow user to edit the feature.

![image](https://user-images.githubusercontent.com/1454356/49645119-b0735700-fa40-11e8-91e6-aa56c736d30b.png)

While analysing this we came across below queries, please let us know how you would like to handle it.

**Case1:**
- Consider there are 2 different features, one created by signed in user(creator of the feature) to which user can edit it, and other created by non-signed in user to which user cannot edit it.
- Now user does multiple selections and clicks on this 2 features.
- As per the current implementation popup tab should be opened in the edit mode displaying similar data of both the features, however, in this case since the user cannot edit other feature, this approach won't work.
- Hence, we are planning to display a message in the popup tab stating "You are not the creator for few of the selected features. So, features could not be edited. Please select only the features created by you."

**Case2:**
- While analysing "ownershipBasedAccessControlForFeatures" property of the layer which only gets filled when "Editors can only edit their features" option is enabled, we observed that it contains property like "allowAnonymousToUpdate".
- Initially when the application is in anonymous mode i.e; the user has not signed in, he is able to edit the feature anonymously to any layer.
- Now, consider in this mode user has selected such a layer to which "Editors can only edit their features" option is enabled and "ownershipBasedAccessControlForFeatures" property is available.
- In this case, if the user tries to edit the feature, we were planning to check the value of "allowAnonymousToUpdate" property. If it is set to true, the anonymous user would be able to edit the feature. If its false, we would provide an error message "As an anonymous user you do not have a permission to edit this feature."

Please let us know if you would like to change any of the above error messages.

**Comment by pratikg** [comment]
@allisonmuise,

> I believe we should already be checking if the current user has the ability to edit the layers in the app on app load.

Yes, as per GitHub ticket #207 we are checking whether Authenticated AGOL user is above level 1 and has editing capability. Only this type of user can edit the feature. For other Authenticated AGOL user who does not fulfil this condition, we throw an error message on click of edit icon "You do not have permission to perform this action."

> Only layers that are editable by the current user should load in the app. (If we aren't, I'll need to add a new issue for that!).

Currently this check is not been added in the code. Should we go ahead and implement this functionality? If Yes, should we completely ignore this type of layer or we should we consider it as a non-editable layer and load it when "showNonEditableLayers" is set to true.

> This will cover the following circumstances where the app shouldn't load at all:
> - If I'm signed in with an account that doesn't have editing capabilities and the app isn't public
> - If the app isn't public and I'm signed in with an account that doesn't have access to the essential apps bundle (test added by Mike T last release)

We need to implement this functionality. We are planning to check this on the application load, if user doesn't have editing capability and is below level 1, application won't load and error message will be displayed in the center of the screen.


> Case1:
> How do we currently handle other errors in saving?

Currently, while editing the popup feature, if it gets failed, we display an error message generated by the "applyEdits"(an operation that edits the feature on the layer) in the alert box and also on the top of the popup form "Report could not be submitted" message is displayed.

> Case1:
> I would suggest only displaying this message when they save their edits.
> I would add the following to that existing error messaging: "Only features created by {username} have been updated." if some features could be updated but not others.

From this we understand that, if user has clicked on multiple features(one created by signed in user(creator of the feature) to which user can edit it, and other created by non-signed in user to which user cannot edit it), and clicks on save button, we will update only those features to which signed-in user can edit and display the message "Only features created by {username} have been updated." Please confirm?

> Case 2:
> These two properties seem to directly contradict one another. Is is actually possible for a user to configure a layer so that users can only edit their features (which requires users to be authenticated in order to track editors) AND still allow anonymous users to edit everything?

For this we have created below cases.

![image](https://user-images.githubusercontent.com/1454356/49732811-c7ff4980-fca5-11e8-86d1-c29df7566970.png)


Please let us know if this answers your question?

**Comment by allisonmuise** [comment]
@pratikg I will take a deeper look at this tomorrow and respond to your questions then.

**Comment by allisonmuise** [comment]
@pratikg 
After taking to some devs and digging into the settings that are now available for layers, here is the suggested set of tests/decisions:

Is editor tracking enabled on the layer (does it track who created features)? 
- If no: The existing checks and behaviors are fine, as all additional checks require editor tracking.
- If yes: Is the current user anonymous?
    - If yes: Do anonymous editors have the same editing privileges as named editors?
        - If no: Only allow them to create related records
        - If yes: Treat them as named editors and proceed to the next bullet point
    - If no: Can editors see all features?
        - if no: only display features created by the current user. Anonymous users with same editing privilege as named editors won't see any features
        - if yes: Can editors edit all features?
            - if no: editors can only edit features they created and add related records. Anonymous users with same editing privilege as named editors will only be able to add related records.
            - if yes: the current behavior is expected.

Level 1 and level 2 are no longer permission levels. The user model released at 6.4 is all capability and role based. Accessing non-public configurable apps requires the user to have access to the essential apps bundle, and their account must have editing capabilities. Mike T added these checks to all our apps last release.

The current error messages and reactions when users can't edit are good - please don't change these based on my earlier comments. I think all we need to do is add in the checks outlined in the tree above a the feature level to fine-tune when the editing button can be used and when related records can be added.

I like your proposed approach to handling batch updates where the user cannot edit all selected features. I assume that this message will accompany the number of feature that were/were not successfully updated as well?

**Comment by aparveen** [comment]
@CTLocalGovTeam While I was testing this issue, I found another issue where if the layer setting is set to "users can only see their own features", Crowdsource Manager fails to load/show any features. 

**Repro Steps:**

1. Create a hosted feature layer.

2. Layer should have the following settings as the screenshot

![image](https://user-images.githubusercontent.com/10148401/53600255-f29ec700-3b5d-11e9-81bc-52e628f06a78.png)

3. Use two separate accounts to add features in the layer.
4. Share the map and the layer with a group.
5. Open the group page and share it to create a Crowdsource Manager app.
6. Configure the default values and launch the app in browser, try to view the app using both account and notice that no features is visible in the app.

![image](https://user-images.githubusercontent.com/10148401/53600643-e7986680-3b5e-11e9-88a2-b2c94f471670.png)

**Comment by sumitzarkar** [comment]
@aparveen 

We are analyzing this issue and looking for a possible solution. This seems to be only reproducible in the devExt. So it might be due to some changes, however we have not finalized on the conclusion.
Once we complete our analysis and found the solution we will update you.

FYI @allisonmuise
